exchange/utils.go

```diff
@@ -700,3 +701,42 @@ func getExtBidAdjustmentFactors(requestExt *openrtb_ext.ExtRequest) map[string]f
 	}
 	return bidAdjustmentFactors
 }
+
+func buildXPrebidHeader(bidRequest *openrtb2.BidRequest, revision string) string {
+	req := &openrtb_ext.RequestWrapper{BidRequest: bidRequest}
+
+	pbsRecord := createNameVersionRecord("pbs-go", revision)
+
+	var channelVersionRecord string
+	if reqExt, _ := req.GetRequestExt(); reqExt != nil {
+		if prebidExt := reqExt.GetPrebid(); prebidExt != nil {
+			if channel := prebidExt.Channel; channel != nil {
+				channelVersionRecord = createNameVersionRecord(channel.Name, channel.Version)
+			}
+		}
+	}
+	var sdkVersionRecord string
+	if appExt, _ := req.GetAppExt(); appExt != nil {
+		if prebidExt := appExt.GetPrebid(); prebidExt != nil {
+			sdkVersionRecord = createNameVersionRecord(prebidExt.Source, prebidExt.Version)
+		}
+	}
+
+	var nonEmptyRecords []string
+	for _, record := range []string{pbsRecord, channelVersionRecord, sdkVersionRecord} {
+		if record != "" {
+			nonEmptyRecords = append(nonEmptyRecords, record)
+		}
+	}
```

Comment:
I like how this code reads, but the creation of a new slices causes two heap memory allocation. We can avoid this using direct checks for each record with a string builder.

Reply:
Agreed, fixed.

Reply:
That's a great suggestion, will do, thanks!